### ISSUE TITLE
OpenGL state manager

### ISSUE THREAD

**Issue by vlidholt** [cocos2d/cocos2d-objc#359]
OpenGL state manager moved from Cocos3d.

**Comment by slembcke** [comment]
So there was a fair amount of unresolved discussion about this. I guess at some point the "reply all" button was missed, but here is the transcript:
http://files.slembcke.net/temp/OpenGLState.txt

The summary of the points I was trying to make is that:
- Caching the state doesn't help with making a threaded renderer.
- Caching the state doesn't help with adding batching.
- Using a wrapped GL API doesn't make writing GL code any easier, and it means that you actually have to "port" it.

While it does make the GL state caching that already exists in Cocos2D more consistent, I'm proposing to replace that with state objects similar to Unity's materials. They are easier to work with in code, thread friendly and are a step in the direction of adding automatic batching.

**Comment by slembcke** [comment]
Bill mentioned that he was starting on this today in the Scrum emails although there has been a lot of unresolved discussion about it.

I _really_ think we should discuss state objects instead of simple caching. I really don't think the latter is going to allow to later improvements of the renderer by means of threading or batching. There have already been a number of people that have expressed disappointment over the lack of an improved renderer in Cocos2D for the 3.0 release.

**Comment by billhollings** [comment]
My main interest in the GL state management is that it needs to be architected and consistent. The historic cocos2d approach of throwing in an arbitrary state-management C-function from release to release has to stop.

Any time new GL state-management is introduced in a release, it is guaranteed to break any sophisticated user's code in unexpected and untraceable ways. Does a user call glXXX directly, or do they now call a new C-function hidden somewhere in the code base?. All of a sudden, a user's custom CCNode will cause ALL cocos2d nodes to simply stop drawing, or to draw strangely, because they're calling glXXX directly somewhere instead of the new state-management function, and thereby breaking the GL state. It often ends up breaking everything that is visible.

From that perspective, we should decide early about how we want to handle state management. And it needs to be an all-or-nothing venture. We should either manage all GL state, or none at all. Otherwise we're in for a long history of breaking user code as each new state management piece is introduced.

I agree with you that cocos2d's use of simple static variables to track state makes it impossible to track state across multiple threads, even loading assets in the background.

The cocos3d approach is to manage all GL state, and at the same time, turn the GL engine into an Objective-C object, rather than a bunch of C-functions. This addresses the threading issue, because these state objects can be assigned on a per-thread basis.

Regarding creating a multi-threaded renderer, I don't consider it to be at odds with this approach. A multi-threaded renderer would batch nodes, or drawing operations, etc, and accumulate them into a queue that would be submitted to the GL state-management system in the serialized order that the GL engine requires. The GL state-management system would not be concerned with ordering the nodes and operations. Its function is simply to make sure the GL engine is not hit with a bunch of redundant GL calls, regardless of how the drawing operations have been batched and submitted by the higher-layer, multi-threaded renderer. I think of the two ideas as two consecutive API layers, with the rendering layer feeding the GL state management layer.

In looking at the current cocos2d code, I think we can start by moving the cocos3d GL tracking framework over to the cocos2d code base, and then redirect all the ccXXX state functions to the equivalent method on CC3OpenGL.sharedGL. This will require no other changes in the cocos2d code base, and doing that much will ensure the following:
- cocos2d & cocos3d manage state consistently
- GL state is tracked separately for the rendering thread and the background thread

This does not solve your valid point about breaking the glXXX design pattern and forcing developers to port their existing code. But I wonder how much that really applies to the cocos2d3d environment. Are we in the position where developers will be porting huge swaths of glXXX code into cocos2d/3d unchanged?

But if that is desired, we'd need to put wrappers around all the glXXX library calls, and in any case, need some kind of internal object environment to be able to delegate to the correct GL state-management code per thread. This can still be compatible with the cocos3d state framework.

In the end, I ask that we either port the cocos3d state framework, or deem the existing state management as arbitrary, rip it out completely, so that the environment is at least consistent, and postpone this issue to a later date.

**Comment by slembcke** [comment]
edit: looks like Bill saw this just as I posted, removed the quote from email since he pasted it above.

It’s definitely a problem that the current GL cache code is incomplete. I’m even partially responsible for a couple of the "arbitrary state-management C-function from release to release” issues since I sent in patches as I ran into missing functionality. I definitely agree that we need something complete, where you don’t touch the GL state directly (pretty much ever). If we are going to do anything to the renderer, this is also what I would consider to be the biggest priority. I’ve had the same problems with having to rewrite custom draw methods every few months.

That said, I’m not specifically opposed to fixing the current state caching, but I don’t think that simple GL state caching is sufficient for future enhancements. To write custom draw methods, you still need to know how GL works even if it’s wrapped up in an Obj-C object since the API exposed is still basically a 1:1 wrapper. If you instead wrap the GL state in discrete objects, you can still perform efficient delta updates to the GL state, and they are much easier to configure and handle. (Again, as a reference, see Unity material objects) They could even be added as a power feature to SpriteBuilder feature to edit them graphically. Unity has it's ShaderLab language for the same purpose, but a GUI editor would make it even easier for beginners to get their feet wet. Custom drawing becomes as simple as configuring a state object and making an array of triangles.

As for threading/batching. I don’t see per thread GL state caches helping. Maybe I’m missing something, but you can’t just run the run the draw methods on a separate thread from the main game loop without running into race conditions. While it would be possible to rewrite all of the built in classes to buffer anything that the draw method needs to read, it would still require locking to write to them (position, blend mode, etc), and it would make custom draw methods really hard to write correctly. On top of that, per thread locking would certainly be slower than an unthreaded renderer, and a global lock would still make it so that only one thread can execute at a time. A lot of Cocos developers want something easy, and a lot of them struggle with writing or using custom GL code. Adding thread locking to that would just make it worse.

What I’m saying is that by buffering the OpenGL state in discrete objects, it makes it possible to buffer the whole draw call. Also, since the renderer can trivially check if two or more sequential draw calls are made with the same state, it makes it possible to batch them by concatenating their vertex data. This is a very common use case in Cocos2D and it becomes very low hanging fruit to implement later if we have state objects.

I'm freaking out a little about this because I don't yet have time to throw together an alternate example implementation, and it's so close to the release date. Your state management code will only make things better than it is now. (I'm not trying to dispute that) Though to really improve the renderer, I think we need to go further. There was some talk about pushing out some major rendering changes to a minor version, even at the cost of annoying a small portion of the users. I'd hate to do it twice in the same major version.
- Scott

**Comment by slembcke** [comment]
<blockquote>
This does not solve your valid point about breaking the glXXX design pattern and forcing developers to port their existing code. But I wonder how much that really applies to the cocos2d3d environment. Are we in the position where developers will be porting huge swaths of glXXX code into cocos2d/3d unchanged?
<br/>
But if that is desired, we'd need to put wrappers around all the glXXX library calls, and in any case, need some kind of internal object environment to be able to delegate to the correct GL state-management code per thread. This can still be compatible with the cocos3d state framework.
</blockquote>

Admittedly the "porting" of GL code I was talking about is sort of rare. I've done it a few times when I prototype ideas outside of Cocos. That's why I added it in as a bit of an afterthought. I mentioned it would be easy to add to the render queue idea using a block and an expensive, but recommendably rare, GL state reset.

<blockquote>
In the end, I ask that we either port the cocos3d state framework, or deem the existing state management as arbitrary, rip it out completely, so that the environment is at least consistent, and postpone this issue to a later date.
</blockquote>

Maybe an important question is how much does the state caching help with performance? I've never actually checked. IIRC, it's optional and not even enabled when running in debug mode right? If it doesn't actually help, I wouldn't mind just going back to the 1.x days where there was no state caching, because I agree that the partial coverage is very frustrating.

I've have used the batching/threading ideas together, and I've used the batching/state object ideas together, and they work very well. BUT I've never used all three together in the same place while trying to make them work in an existing engine. I definitely want to take a couple of days to hack an example together to make sure it works as well as I think it does within Cocos.

How annoyed would users be if we released 3.1 with the improved renderer even if introduced breaking changes to custom drawing? Would it be worthwhile to make a forum thread about this?

**Comment by billhollings** [comment]
@slembcke 

Can you educate me on how Unity's concept of a Material differs from cocos2d's concept of a drawable CCNode (sprite, button, etc)?

I'm asking because in cocos3d, the equivalent of a drawable node is CC3MeshNode, which is basically a CC3Mesh and a CC3Material, and a CC3Material is configurable colors, textures and shaders:
- CC3MeshNode
  - CC3Mesh
    - vertex attributes
  - CC3Material
    - CC3Textures...
    - CC3ShaderProgram
    - colors, opacity, shininess, etc

So...in my world view:
- Your visual editor would spit out a CC3Material and CC3ShaderProgram, which would then be attached to a CC3MeshNode (or multiple mesh nodes).
- Your (possibly multi-threaded) RenderQueue would then:
  - queue CC3MeshNodes for rendering
  - order them according to Z-order rules for transparency
  - combine the meshes of CC3MeshNodes containing the same opaque CC3Materials (batching)
  - submit them to the GL state-manager:

And the layered flow of each render loop becomes:

Scene -> RenderQueue -> GLStateManager -> glXXX...

Does this layering make sense? If so...what we're addressing in this thread, and what is being proposed, is just the GLStateManager layer. It does not preclude the development of the RenderQueue.

BTW...this is the current behaviour of cocos3d. It already has an active render queue that flattens and orders the CC3MeshNodes according to Z-order, and it has a GL state manager. The flow is as described above. The filling of the render queue is not multi-threaded, and batching is not automatic yet, but both of those features can be added at some point. But again, what we're talking about here is just the GLStateManager layer.

**Comment by billhollings** [comment]
The other nice thing the current cocos3d state tracker does is support optional GL error and tracing logging. Which is really handy for debugging GL state. It would be very useful if cocos2d supported this as well. :^)

**Comment by billhollings** [comment]
However...as I say, if all of this is too much design to address at this point, I'd be satisfied if we just ripped the existing cocos2d state management out altogether.

**Comment by vlidholt** [comment]
I'm no expert on this. But to me it sounds like a win as long as it doesn't interfere with the plans of adding support for the automatic batch rendering.

Also, for potentially - in the future - adding a threaded renderer, which tasks can be done in parallel?

Btw, has anyone had a look at the cocos2d-x implementation?

**Comment by Birkemose** [comment]
I think we should be careful not rushing new functionality into cocos2d. For v3, the old state manager should be removed though.

I also think we are barking up the wrong tree here. Cocos2d is not slow as it is now, and if we scrap the state manager, it is also not difficult to customize. That is what has made it so popular. So if we want to improve it, a state manager, or a gl wrapper will make no difference in the end.

If we want to make a fast 2d renderer, we first of all has to forget everything about 3d, and start thinking out of the box. 

Commands to the central renderer must obviously be cashed, otherwise we have no way of manipulating them later. It also gives the advance of being able to add more advanced commands in updates. It should be simple and highly flexible command system. I have extensive experience in building such systems.

I have tested various ways of switching textures in the vertex shader, and while it works fine, it only has it's advantages, when many small sprites are rendered from many different textures, which is a highly unlikely scenario.

So as for textures, the only real way of increasing performance, is to generate the sprite sheets on the fly, for minimal texture switches. I would guess, that this drastically could reduce number of batches, plus it would add an unpresidented flexibility to both artwork layout, and ease of use. 
That would be true next gen.

I have a lot of other ideas for a next gen renderer, like the nodes converting their vertices directly into vbo. Enabling hardware z buffer. Automatic geometry breakdown. Sorting by using indexed rendering. Large scale, single node, non sorted particle systems. Tabeled transformations. Multifunction shaders. 

So please let us not hurry anything into first release.

**Comment by Birkemose** [comment]
As for threaded rendering, I would like to see some hard evidence, that it is worth it, and "Quake uses it", will not suffice. As the GPU of mobile devices is fairly weak, I can not imagine a scenario (even in non vbo 2.x) where the CPU is the bottleneck. If that is so, it makes no sense to break the render engine in two.

**Comment by billhollings** [comment]
Okay. 'Nuff said. 

I'll back off porting any of the cocos3d GL state management over.

Who's wants to take care of disabling the existing cocos2d GL state? 

The trivial and expedient thing to do would be to hardcode CC_ENABLE_GL_STATE_CACHE to 0 in ccConfig.h (and removing the #ifdef, plus some commentary about how it is deprecated). 

But that would still leave the weird half-API. It's probably best to strip it right out, and replace the proper glXXX calls back into the library.

Does anyone have an inkling to do that as part of some other related task, or should I take care of it?

**Comment by FramusRock** [comment]
To me it sounds really good what slembcke and bill say. For me the thing I was most anticipating for the new version is a new (faster) renderer.

I'd really plan anything that needs to be done to the new rendering first since it's the most important part of a game / graphics engine. Is the existing rendering code base really that good? Let me make a tiny example: cocos2d uses lines like these in various places

```
glVertexAttribPointer(kCCVertexAttrib_Color, 4, GL_UNSIGNED_BYTE, GL_TRUE, kQuadSize, (GLvoid*) offsetof( ccV3F_C4B_T2F, colors));
```

where GL_FLOAT is faster as far as I know since it gets converted to float later anyway. I really don't want to go offtopic with that. I'd rather like to encourage somebody that is an expert at OpenGL to look over the code and re-judge anything there from today's point of view.

Then I'd think about anything else e.g. how we can go multi-threaded. It's really a pity that cocos2d does not benefit from all cores of today's devices and I think there's a lot of potential. The thread locking and syncing can be some hard work to figure out but it's probably worth it in the end if done at the right places. Of course these places are to be discussed and then we can profile it in every situation and find some evidence that it's indeed worth it. 

@slembcke I'm totally with you and I'd love to see your prototype and how you can make the three things work together but is all that not dependant on the underlying code that potentially could be improved? Please correct me if I'm wrong on this!

**Comment by slembcke** [comment]
Bwa. Lots of replies. This might be a long post.

@billhollings So in Unity, a Material is a set of rendering passes (though most have only one), and each pass is a more or less full GL state (blending mode, z testing mode, shader + variables, etc) except for the vertex bindings. Unity provides a GL-like custom rendering API as well as the Graphics.DrawMesh() method. You control the graphics state by binding a material pass first, then making drawing commands. When doing regular rendering, a GameObject may have a Mesh associated with it (via a MeshFilter component) and a set of materials (via a Renderer component). You need both in order to get something to draw on the screen. I guess the reason for the MeshFilters and Renderer components being separate is ancient history predating vertex shaders. After taking a look at your material class, it's fairly comparable to Unity's. Here are some references if you are interested:

Shaders are defined using the "ShaderLab" language. It's hard to find good complete examples... but here is the overall docs for it. It's a combination of blending modes, z-testing, a shader program (Cg, GLSL, etc), shader variables, etc.
http://docs.unity3d.com/Documentation/Components/SL-Reference.html

The API is mostly about managing shader parameters. You can't change the blending mode without more ShaderLab code for instance.
http://docs.unity3d.com/Documentation/ScriptReference/Material.html

Basically what I'm proposing is to make something that is somewhere between your state cache and material class. The renderer maintains a current state object that matches GL's, and when setting a new state it can efficiently perform a delta update of the GL state. By allowing one to clone the current or any existing state object, you can easily allow for anything that you could have done with the raw GL functions, but in a much easier to learn package. The state objects can also be buffered at any time (even easier if we make them immutable) to make implementing threading easy and safe.

Again, the issue I have with queueing nodes instead of drawing commands is that nodes cannot be buffered for a secondary rendering thread, while drawing commands can. You only need to buffer the state object and the vertex data for a draw. Passing entire nodes to a secondary thread would require locking the entire scene while rendering.

@Birkemose There are two big advantages of threaded rendering. The first is that it doesn't block the main thread. While the GPU is waiting for vsync, your main thread will be stuck waiting and unable to do anything. This is particularly bad if frames are taking slightly more than 16ms per frame. You'll spend half of your time blocked and waiting for vsync. This is useful even on the old iPhone 4 with a single CPU. The other big advantage is that presenting the render buffer and making the draw calls can use a significant amount of CPU time. On many operating systems, if your main thread is doing too much work and interfering with the normal event loop it can cause _extremely_ severe input lag. On iOS, this is often several seconds. This is a fairly common question on the support forums. I can make an example with the Chipmunk Showcase app that demonstrates both issues. (perhaps after 3.0 ships) I'm fairly certain (though guessing) that threading the renderer using a queue is the single biggest "free" optimization we can make. What I mean by "free" is that it will make things faster without requiring users to do anything to take advantage of it.

@FramusRock The GPU will convert it to a float eventually and that conversion may or may not be expensive (no idea really). It can save a lot of memory and memory bandwidth to use a lower precision type. That said, for 2D games, the vertex data is _rarely_ even close to being a bottleneck. Using floats instead of fixed precision color formats has a lot of other nice advantages too.

I think I hit all of the important parts?

**Comment by cocojoe** [comment]
@slembcke That all sounds good.

On 25 November 2013 19:08, slembcke notifications@github.com wrote:

> Bwa. Lots of replies. This might be a long post.
> 
> @billhollings https://github.com/billhollings So in Unity, a Material
> is a set of rendering passes (though most have only one), and each pass is
> a more or less full GL state (blending mode, z testing mode, shader +
> variables, etc) except for the vertex bindings. Unity provides a GL-like
> custom rendering API as well as the Graphics.DrawMesh() method. You control
> the graphics state by binding a material pass first, then making drawing
> commands. When doing regular rendering, a GameObject may have a Mesh
> associated with it (via a MeshFilter component) and a set of materials (via
> a Renderer component). You need both in order to get something to draw on
> the screen. I guess the reason for the MeshFilters and Renderer components
> being separate is ancient history predating vertex shaders. After taking a
> look at your material class, it's fairly comparable to Unity's. Here are
> some references if you are interested:
> 
> Shaders are defined using the "ShaderLab" language. It's hard to find good
> complete examples... but here is the overall docs for it. It's a
> combination of blending modes, z-testing, a shader program (Cg, GLSL, etc),
> shader variables, etc.
> http://docs.unity3d.com/Documentation/Components/SL-Reference.html
> 
> The API is mostly about managing shader parameters. You can't change the
> blending mode without more ShaderLab code for instance.
> http://docs.unity3d.com/Documentation/ScriptReference/Material.html
> 
> Basically what I'm proposing is to make something that is somewhere
> between your state cache and material class. The renderer maintains a
> current state object that matches GL's, and when setting a new state it can
> efficiently perform a delta update of the GL state. By allowing one to
> clone the current or any existing state object, you can easily allow for
> anything that you could have done with the raw GL functions, but in a much
> easier to learn package. The state objects can also be buffered at any time
> (even easier if we make them immutable) to make implementing threading easy
> and safe.
> 
> Again, the issue I have with queueing nodes instead of drawing commands is
> that nodes cannot be buffered for a secondary rendering thread, while
> drawing commands can. You only need to buffer the state object and the
> vertex data for a draw. Passing entire nodes to a secondary thread would
> require locking.
> 
> @Birkemose https://github.com/Birkemose There are two big advantages of
> threaded rendering. The first is that it doesn't block the main thread.
> While the GPU is waiting for vsync, your main thread will be stuck waiting
> and unable to do anything. This is particularly bad if frames are taking
> slightly more than 16ms per frame. You'll spend half of your time blocked
> and waiting for vsync. This is useful even on the old iPhone 4 with a
> single CPU. The other big advantage is that presenting the render buffer
> and making the draw calls can use a significant amount of CPU time. On many
> operating systems, if your main thread is doing too much work and
> interfering with the normal event loop it can cause _extremely_ severe
> input lag. On iOS, this is often several seconds. This is a fairly common
> question on the support forums. I can make an example with the Chipmunk
> Showcase app that demonstrates both issues. (perhaps after 3.0 ships) I'm
> fairly cert ain (though guessing) that threading the renderer using a queue
> is the single biggest "free" optimization we can make. What I mean by
> "free" is that it will make things faster without requiring users to do
> anything to take advantage of it.
> 
> I think I hit all of the important parts?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/cocos2d/cocos2d-iphone/issues/359#issuecomment-29231609
> .

## 

Martin

**Comment by slembcke** [comment]
I should restate so that it's clear. I agree with Bill that we should hold off on any major rendering changes until after the 3.0 release. There just isn't enough time to work out both what Cocos2D will need in a new renderer (which several people have competing ideas for, but none have started) and what Cocos3D needs. It's way to close to the release date to make any major changes.

**Comment by FramusRock** [comment]
@slembcke That sounds really good! The blocking and waiting for vsync happened so often to me and it also showed up in the support forums several times (It even caused memory problems to some people). I modified cocos2d in my builds, so that it skips frames if a frame took longer than 16 ms to calculate otherwise there's big trouble going on. So I totally agree with you that this will not only help on the newer device but also on the old iPhone 4 it will be nice enhancement.

About the color types: I think we should take every "free" optimization (as you said) that we can get. And I think it would be a good idea create a (padded) vertex struct with all floats and use it throughout cocos2d (or at least at the places where Byte color types are used). With saying that cocos2d still uses non-float color format I just wanted to make an example. I think there are many tiny things like this that add up to take GPU or CPU time and that could potentially be improved. Going through the underlying rendering code and profiling / fixing   many small things can add up to a big improvement in the end.

Edit: slembcke too fast ;) Yes, I'd love to see a good and well-thoughtout renderer change. If it's too close now, let's face it as soon as possible after 3.0 is out.

**Comment by slembcke** [comment]
Hmm. Shoot. My memory of the state cache performance may have been a little too specific to what I was making. (A few large custom draw things that used a lot of fillrate)

I just ran the sprite test now with and without the cache enabled. The A9 test (a few dozen unbatched pokemon images) with 2000 onscreen sprites on an iPad 2 went from 9 to 17 fps when enabling the cache. The A1 test with 2000 of the same sprite went from 9 to 22 fps. That is perhaps not an average scene, but still pretty important results. :-\

So maybe for the 3.0 release the best course of action is still to disable the state cache by default and deprecate it as Bill proposed above, but leave the functions in for now (until 3.1?) for compatibility with 2.x custom drawing code while we decide what to do.

The other question which has come up recently is what rendering changes should happen in 3.1 vs 4.0. In my opinion, we should really focus on an overhaul of the renderer for 3.1. It's too big to ignore even if it's going to break backwards compatibility and not strictly be semantic versioning. At least by leaving 3.0 as is, we won't be breaking custom rendering code twice. There is a lot of low hanging fruit that can be improved, and I would hate to wait until 4.0 to do anything significant.

**Comment by Panajev** [comment]
Uhm,

Maybe I am misunderstanding you here but...

"I just ran the sprite test now with and without the cache enabled. The A9
test (a few dozen unbatched pokemon images) with 2000 onscreen sprites on
an iPad 2 went from 9 to 17 fps when enabling the cache. The A1 test with
2000 of the same sprite went from 9 to 22 fps. That is perhaps not an
average scene, but still pretty important results. :-\ So maybe for the 3.0
release the best course of action is to disable the state cache by default
and deprecate it as Bill proposed above. "

I read that as "caching massively improve performance" --> "cache should be
disabled by default and removed" ?

**Comment by slembcke** [comment]
Well yes. That's sort of the problem. The caching is problematic since it's very incomplete. In past projects where I've written significant custom drawing, disabling the caching didn't make much of a difference to performance at all. Since it's easier to avoid the Cocos GL cache functions, I was perfectly happy with the cache disabled.

In scene with a lot of small sprites, it does make a _huge_ difference. Disabling it by default and deprecating it allows us to start moving forward with even bigger improvements. The functionality will still be there if somebody really needs to enable it (vs. removing it entirely). I agree it's a rather imperfect sounding short term solution though.

**Comment by billhollings** [comment]
@slembcke 

Thanks for the Unity intel !

And thanks for running those tests. Nice to see that we are not barking up the wrong tree ! :^)

From what I see then...a queue of some kind of encapsulated drawing state, that can possibly be pulled from a different thread...plus some kind of GL state management would be the best arrangement. I still see that as some kind of layered API, as I described above.

However...to be honest...I'm still not sure how much the secondary queue of drawing objects will help. If the CPU filling the queue is the bottleneck (because one is running some big-ass re-calculation on the scene...the several seconds that you describe above), then the drawing thread will stall once it has drained the queue. Isn't it just postponing the problem?

From my thinking, it's the thread that is _filling_ the hypothetical queue that needs to be kept unencumbered...the one running the CCActions, etc. Might other threads might be better used running physics, etc, in the background? However...in the edge cases that you describe, where grabbing the chunks of time around the buffer swapping and vsyncs can make a difference, I can see that it possibly might help.

**Comment by slembcke** [comment]
Oh. No no. To be more clear, the latency is because if you come close to 100% CPU utilization on the main thread, the OS starts buffering input events. I've seen very similar things happening on OS X and Windows too. Overtaxing the main thread is not an iOS specific issue at all, though it tends to be more severe.

If you look at time profiler traces of common Cocos2D programs, GL uses a _huge_ amount of the CPU time. IIRC, between glDrawArray() and presentRenderbuffer it's usually almost half of the CPU time spent on the main thread. Those are the parts you want to push off to a secondary thread.

Filling the render queue itself is very cheap (for 2D sprites anyway). Just some memcpy() calls, or transforming geometry directly into double buffered VBO. Certainly a lot cheaper than submitting the actual draw call to GL.

**Comment by Panajev** [comment]
Thanks for the update :). Overall, I agree that the revamp would bring a lot of benefits in post 3.0 releases to afford that kind of change now. Companies and individuals falling in your first example would not be hurt, but potentially lots of other users of cocos2d might, users without lots of custom drawing methods.

Also, and yes I am partial to it, something the old cache was doing was uniform caching, trading a User CPU time hit vs a potentially quite bigger GL driver CPU hit when you were updating a uniform of a shader object with the same value every frame. Maybe it is not a concern to many, but in the past I have seen quite a bit of a speed up by doing that vs letting the GL driver  do better. Thoughts on that front?

[web [iPhone msg]]

**Comment by FramusRock** [comment]
@slembcke Totally agree, this is exactly what my profiling results were. I've also seen that huge amount of CPU being used at that place.

**Comment by Birkemose** [comment]
Well, it seems that consensus is, that adding threading, and a solid state handling, will be enough to bring the render engine up to date.
I strongly disagree on that, but if I am the only one with that opinion, I will rest my case.

**Comment by slembcke** [comment]
I have a few goals I'd like to see improve that the state objects help with:
- Writing custom drawing code requires knowing too much about GL. 90% of the time it should be as simple as setting up the drawing state and making an array of triangle data. That's much easier to explain than GL.
- It should be threadable. (I have some strong examples where this is very beneficial)
- It should be batchable.
- Shaders need to be easier to set up.

I'm not opposed to the fancier features you mentioned, I'm only advocating that a simple state cache won't be sufficient to build fancier features upon later.

**Comment by slembcke** [comment]
I have a few goals I'd like to see improve that the state objects help with:
- Writing custom drawing code requires knowing too much about GL. 90% of the time it should be as simple as setting up the drawing state and making an array of triangle data. That's much easier to explain than GL.
- It should be threadable. (I have some strong examples where this is very beneficial)
- It should be batchable.
- Shaders need to be easier to set up.

~~I'm not opposed to the fancier features you mentioned, I'm only advocating that a simple state cache won't be sufficient to build fancier features upon later.~~

edit: That last bit came off a bit pointed perhaps. Nobody has yet been able to agree on how to move forward, so I'm just agreeing that we shouldn't rush into anything since there isn't enough time to really figure how to get everybody what they are looking for.

**Comment by FramusRock** [comment]
Maybe we should open another thread and "collect" all the ideas and think about how we put them together to make the render engine better.

edit: This means the ideas you all mentioned sound good but written down in a nice way makes it easier to get an overview.
